### PR TITLE
Operate keyword

### PR DESCRIPTION
**Pull request by joakim-hove**

### CODE REVIEW

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp [hunk 1]
```diff
@@ -397,6 +398,118 @@ namespace Opm {
...
+        double COPY(double, double X, double, double ) {
+            return X;
+        }
+
+        double MAXLIM(double, double X, double alpha, double ) {
+            return std::min( alpha , X );
+        }
+
+        double MINLIM(double, double X, double alpha, double ) {
+            return std::max( alpha , X );
```
Comment (pgdr) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: Should really `MINLIM` use `max` and `MAXLIM` use `min`?

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp [hunk 2]
```diff
@@ -397,6 +398,118 @@ namespace Opm {
...
+        const std::string& targetArray = record.getItem("TARGET_ARRAY").get< std::string >(0);
+        const std::string& operation   = record.getItem("OPERATION").get< std::string >(0);
+        double alpha = record.getItem("PARAM1").get< double >(0);
+        double beta = record.getItem("PARAM2").get< double >(0);
+
+        if (!supportsKeyword( targetArray))
+            throw std::invalid_argument("Fatal error processing COPYREG record - invalid/undefined keyword: " + targetArray);
+
+        if (!hasKeyword( srcArray ))
+            throw std::invalid_argument("Fatal error processing COPYREG record - invalid/undefined keyword: " + srcArray);
```
Comment (pgdr) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: `COPYREG` should be `OPERATE`?

File: opm/parser/eclipse/EclipseState/Grid/GridProperty.hpp [hunk 3]
```diff
@@ -103,7 +103,9 @@ class GridProperty {
     void iset(size_t index, T value);
     void iset(size_t i , size_t j , size_t k , T value);
 
+
     const std::vector<T>& getData() const;
+    std::vector<T>& getData();
```
Comment (pgdr) [inline opm/parser/eclipse/EclipseState/Grid/GridProperty.hpp]: why is this introduced now?

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp [hunk 4]
```diff
@@ -397,6 +398,118 @@ namespace Opm {
...
+        }
+
+        // NB: The POLY function and the MULTIPLY function both use
+        // the R value in the calculation. That implies that we should
+        // ideally check that the R property has already been
+        // initialized with a valid value, For all the other
+        // operations R only appears on the left side of the equation,
+        // and can be fully assigned to.
+        double POLY(double R, double X, double alpha, double beta) {
+            return R + alpha * pow(X , beta );
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: The `std::pow` function from cmath was already available here, no?

File: opm/parser/eclipse/EclipseState/Grid/GridProperty.hpp (hunk 3 above)
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Grid/GridProperty.hpp]: Looks like OPERATE can modify grid properties as well, so I'd wager it's to facilitate that.

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp (hunk 2 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: Ehhh - yes.

File: opm/parser/eclipse/EclipseState/Grid/GridProperty.hpp (hunk 3 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Grid/GridProperty.hpp]: > Looks like OPERATE can modify grid properties as well, so I'd wager it's to facilitate that.

Yes that is right - the other keywords which can modify a property, like e.g. `COPY` and `MULT` are implemented with mutating property methods. In this case the required property method had to have a quite weird signature and the implementation is therefor applied directly to the underlying `std::vector<T>`.

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp (hunk 4 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: I guess - I will look into it.

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp (hunk 1 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: Well - at least according to the documentation - yes; I had a glimpse of understanding while developing this - but that frankly eludes me now.

**Comment by pgdr** [comment]
Build failed, but due to time constraints on Travis' side.  I've requested a rebuild.

**Comment by pgdr** [comment]
Again failed caused by a timeout in compiling ewoms.

**Comment by pgdr** [comment]
jenkins build this please

**Pull request merged**